core/platform_feature_list.py

```diff
@@ -53,11 +53,11 @@
 # Names of features in prod stage, the corresponding feature flag instances must
 # be in prod stage otherwise it will cause a test error in the backend test.
 PROD_FEATURES_LIST = [
```

Comment:
Hi @nithinrdy, I checked with Ben on how platform params are supposed to work and got more clarification on this.

Firstly, in terms of workflow: what seems ideal is that there's a way for a feature to be on in develop and off in production, and turned on in production later when Diana (or the PM team) have approved it. (The idea of gating this PR on PM review seems counter to this ideal, so I checked with Ben to see if I was thinking about this workflow correctly.)

Per @BenHenning, this workflow is exactly what is intended and how things should work today if the code is gated by the platform parameter's boolean value. It will only default on for the current build environment of the server (i.e. if it's in 'dev' stage then it's automatically on for the dev environment, but will be off for test & production). If we want to follow that in this case, we would not want to remove the platform parameter in this PR as that will result in the feature being enabled automatically in the next deployment that PR is included in.

For integration tests, https://github.com/oppia/oppia/blob/develop/core/tests/protractor/featureGatingFlow.js is the demo test for showing how to set this up in the protractor environment (it's basically just adding a step to navigate to the admin panel to enable the test, and then running the specific test). Might this help?

Also adding @BenHenning to clarify anything I've missed, or for further follow-ups.

Reply:
Hi Sean, thanks for taking a look. 

> It will only default on for the current build environment of the server (i.e. if it's in 'dev' stage then it's automatically on for the dev environment, but will be off for test & production).

Could you maybe define 'on' and 'off' in this context? Because the way that I noticed the flags function on my local server was a little different.

From what I've noticed on my local dev server (and based on the code in the `create_platform_parameter` method in `platform_parameter_registry.py`), the features are disabled by default regardless of feature stage (i.e. whether the feature flag is labeled to be in the DEV, TEST, or PROD stages, it is disabled on all of them by default; DEPRECATED is different from these of course).

The feature stages don't define the environment in which the featues are turned on, [they define the environments where the features COULD be turned on in](https://github.com/oppia/oppia/wiki/Developing-new-features-with-feature-gating#feature-stage-explanation).

A feature in the DEV stage is disabled by default on all environments, but can be turned on in the DEV environment by an admin. It cannot be turned on in testing and production environments by said admin (this is the stage I put the flag in in [the first PR](https://github.com/oppia/oppia/pull/15589/files), because the feature was incomplete and there was no reason to turn it on in prod).

A feature in the PROD stage is, again, disabled by default in all environments, but can be turned on in dev, test and prod enviromnents by an admin. (this is the stage I put the flag in in [the third PR](https://github.com/oppia/oppia/pull/15634/files), because the feature was now complete and it may now be turned on in prod/test environments for testing).

I found the above behaviour to be consistent across both modes of my local server (i.e. in dev mode and in prod mode).

Unless I'm doing something wrong/feature flags function differently on _oppiatestserver.org_ and _oppia.org_, the above description and rules should apply there too.

(I'd be happy to know if I'm missing something/got something wrong in this comment @seanlip @BenHenning).

However, like you stated, removing the flag will indeed enable the features on the next deployment which includes this PR, but I assumed this _was_ the intention, i.e. to remove the flag and enable the features once we get an approval from Diana (This PR is meant to be merged after Diana tests these features on the test server)?

Is the flag to be removed at some later date i.e. after the evaluation?
Please confirm if you'd like me to leave the flag in PROD stage and I'll rewrite the E2E tests accordingly. Thanks!

Reply:
Thanks @nithinrdy -- great analysis. It is possible that the system is not working as intended, and if that's the case, this PR is indeed blocked on Diana's approval.

But in order to figure out whether this is the case, I think we need confirmation from @BenHenning since he was the one who oversaw platform parameters (and I have limited knowledge about the system, so I'm currently playing messenger). Could you discuss this with him directly, please?

Thanks!

Reply:
Got it @seanlip, I'll talk to Ben regarding this.

Could you please confirm this part of the comment, as it is concerned more with my project and not the gating system, and I'm a little unclear about it?

> However, like you stated, removing the flag will indeed enable the features on the next deployment which includes this PR, but I assumed this _was_ the intention, i.e. to remove the flag and enable the features once we get an approval from Diana (This PR is meant to be merged after Diana tests these features on the test server (which Kevin said should be done in a day or so))?
> 
> Is the flag to be removed at some later date i.e. after the evaluation?

Thanks!

Reply:
Yup you can totally get the rest of the reviews done in the meantime.

Re the quoted comment, what I stated was based on my (incorrect? correct?) understanding of how the feature gating system is supposed to function. My take is that overall, I would prefer if we could merge the e2e tests to develop without being worried about having to tweak the platform parameters; these things should really be independent and what that would mean is that the platform parameter tweaking wouldn't be in this PR. But it seems like these are currently a bit intertwined.

So: if they are intertwined and this isn't easily resolvable, then my earlier comment applies (i.e. Diana approves, I approve removal of the flag, we merge).

And if they are not intertwined, then it would be better if this PR just had the e2e tests and the feature turned on in dev (but not prod), and then we can turn on the feature in prod whenever we want (and there is less time pressure on the feature review due to the "blocking code merging" side effect).

Does that help clarify?

Reply:
Yes, thank that clears things up thank you

Reply:
Hi. I think I was wrong yesterday when I was providing context to you @seanlip. @nithinrdy's description above sounds correct--the stage defines *which* environments the parameter is allowed to be enabled in.

Regarding the question of when to enable this, I think that's up to you @seanlip. The intention would be that once this feature is finished, it's promoted to testing stage for QA verification. After it's approved by QA, it can enabled for production and enabled after the next release goes out. We would only deprecate the feature once we're confident we would not want to ever roll it back.

Reply:
OK, thanks @BenHenning!

In that case, @nithinrdy, let's leave the flag in PROD stage and turn it on after the deployment is made. I think we should start practising the use of this flow for feature flags in general, and this is a good feature to do that for. Would that be OK?

Reply:
Thanks @BenHenning!

@seanlip Okay so I've reverted the changes which correspond to deprecating the flag, and moved it back to the PROD stage.
I've now modified the E2E tests to first enable the feature flag in prod mode before running the topic and story viewer tests.